File: smoothie-sample/src/main/java/com/example/smoothie/RxMVPActivity.java
package com.example.smoothie;

import android.app.Activity;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;
import butterknife.BindView;
import butterknife.ButterKnife;
import com.example.smoothie.deps.RxPresenter;
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;
import javax.inject.Inject;
import rx.Subscription;
import toothpick.Scope;
import toothpick.Toothpick;
import toothpick.smoothie.module.SmoothieActivityModule;

import static toothpick.Toothpick.openScopes;
import static toothpick.smoothie.module.Smoothie.autoCloseActivityScopeAndParentScope;

public class RxMVPActivity extends Activity {

  public static final Class PRESENTER_SCOPE = Presenter.class;
  private Scope scope;

  @Inject RxPresenter rxPresenter;
  @BindView(R.id.title) TextView title;
  @BindView(R.id.subtitle) TextView subTitle;
  @BindView(R.id.hello) Button button;
  private Subscription subscription;

  @Override protected void onCreate(Bundle savedInstanceState) {
    scope = openScopes(getApplication(), PRESENTER_SCOPE, this);
    scope.installModules(new SmoothieActivityModule(this));
    autoCloseActivityScopeAndParentScope(this);

    super.onCreate(savedInstanceState);
    Toothpick.inject(this, scope);

    setContentView(R.layout.simple_activity);
    ButterKnife.bind(this);
    title.setText("MVP");
    subscription = rxPresenter.subscribe(aLong -> subTitle.setText(String.valueOf(aLong)));
    button.setVisibility(View.GONE);
  }

  @Override
  protected void onDestroy() {
    subscription.unsubscribe();
    super.onDestroy();
  }

  @javax.inject.Scope @Target(ElementType.TYPE) @Retention(RetentionPolicy.RUNTIME)
  public @interface Presenter {
  }
}

File: smoothie-sample/src/main/java/com/example/smoothie/kotlin/RxMVPActivity.kt
package com.example.smoothie.kotlin

import android.app.Activity
import android.os.Bundle
import android.view.View
import android.widget.Button
import android.widget.TextView
import com.example.smoothie.R
import com.example.smoothie.kotlin.deps.RxPresenter
import rx.Subscription
import rx.functions.Action1
import toothpick.Scope
import toothpick.Toothpick
import toothpick.kotlin.injectLazy
import toothpick.smoothie.module.SmoothieActivityModule

class RxMVPActivity : Activity() {
    private lateinit var activityScope: Scope

    // Uses lazy assign of scope via InjectConfig block and default TP Kotlin Extension
    private val rxPresenter: RxPresenter by injectLazy { scope = activityScope }

    private val title by lazy { findViewById<TextView>(R.id.title) }
    private val subTitle by lazy { findViewById<TextView>(R.id.subtitle) }
    private val button by lazy { findViewById<Button>(R.id.hello) }

    private lateinit var subscription: Subscription

    override fun onCreate(savedInstanceState: Bundle?) {
        activityScope = Toothpick.openScopes(application, PRESENTER_SCOPE, this)
        activityScope.installModules(SmoothieActivityModule(this))
        super.onCreate(savedInstanceState)

        setContentView(R.layout.simple_activity)
        title.text = "MVP"
        subscription = rxPresenter.subscribe(Action1 { aLong -> subTitle.text = aLong.toString() } )
        button.visibility = View.GONE
    }

    override fun onDestroy() {
        Toothpick.closeScope(this)
        subscription.unsubscribe()
        if (isFinishing) {
            //when we leave the presenter flow,
            //we close its scope
            rxPresenter.stop()
            Toothpick.closeScope(PRESENTER_SCOPE)
        }
        super.onDestroy()
    }

    @Target(AnnotationTarget.CLASS)
    @MustBeDocumented
    @Retention(AnnotationRetention.RUNTIME)
    @javax.inject.Scope
    annotation class Presenter

    companion object {
        val PRESENTER_SCOPE: Class<*> = Presenter::class.java
    }
}
